File: shen-derive/core/sexpr_print.go
package core

import (
	"fmt"
	"strings"
)

// PrettyPrintSexpr renders an s-expression in Shen surface syntax.
func PrettyPrintSexpr(s Sexpr) string {
	if s == nil {
		return "()"
	}
	return ppSexpr(s)
}

func ppSexpr(s Sexpr) string {
	switch s := s.(type) {
	case *Atom:
		if s.Kind == AtomString {
			return fmt.Sprintf("%q", s.Val)
		}
		return s.Val
	case *List:
		if len(s.Elems) == 0 {
			return "()"
		}

		// Try to render as [a b c] sugar for cons chains
		if elems, tail, ok := desugarConsList(s); ok {
			parts := make([]string, len(elems))
			for i, e := range elems {
				parts[i] = ppSexpr(e)
			}
			if IsSym(tail, "nil") {
				return "[" + strings.Join(parts, " ") + "]"
			}
			return "[" + strings.Join(parts, " ") + " | " + ppSexpr(tail) + "]"
		}

		parts := make([]string, len(s.Elems))
		for i, e := range s.Elems {
			parts[i] = ppSexpr(e)
		}
		return "(" + strings.Join(parts, " ") + ")"
	}
	return fmt.Sprintf("%v", s)
}

// desugarConsList checks if a list is a cons chain:
// (cons a (cons b (cons c nil))) → [a, b, c], nil, true
// (cons a (cons b tail)) → [a, b], tail, true
func desugarConsList(l *List) (elems []Sexpr, tail Sexpr, ok bool) {
	if len(l.Elems) != 3 || !IsSym(l.Elems[0], "cons") {
		return nil, nil, false
	}

	elems = append(elems, l.Elems[1])
	rest := l.Elems[2]

	for {
		rl, isList := rest.(*List)
		if !isList || len(rl.Elems) != 3 || !IsSym(rl.Elems[0], "cons") {
			return elems, rest, true
		}
		elems = append(elems, rl.Elems[1])
		rest = rl.Elems[2]
	}
}

File: cmd/shen-derive-ts/core/sexpr-print.ts
// Pretty-printer for s-expressions in Shen surface syntax.
// Ported from shen-derive/core/sexpr_print.go.

import { type Sexpr, type SList, isSym } from "./sexpr.ts";

// prettyPrintSexpr renders an s-expression in Shen surface syntax.
export function prettyPrintSexpr(s: Sexpr): string {
  return ppSexpr(s);
}

function ppSexpr(s: Sexpr): string {
  if (s.kind === "atom") {
    if (s.atomKind === "string") {
      return quoteString(s.val);
    }
    return s.val;
  }

  // list
  if (s.elems.length === 0) {
    return "()";
  }

  // Try to render as [a b c] sugar for cons chains
  const desugared = desugarConsList(s);
  if (desugared !== null) {
    const { elems, tail } = desugared;
    const parts = elems.map(ppSexpr);
    if (isSym(tail, "nil")) {
      return "[" + parts.join(" ") + "]";
    }
    return "[" + parts.join(" ") + " | " + ppSexpr(tail) + "]";
  }

  const parts = s.elems.map(ppSexpr);
  return "(" + parts.join(" ") + ")";
}

// quoteString mirrors Go's fmt.Sprintf("%q", ...) for the common escape cases.
function quoteString(s: string): string {
  let out = '"';
  for (const ch of s) {
    switch (ch) {
      case "\\":
        out += "\\\\";
        break;
      case '"':
        out += '\\"';
        break;
      case "\n":
        out += "\\n";
        break;
      case "\r":
        out += "\\r";
        break;
      case "\t":
        out += "\\t";
        break;
      default: {
        const code = ch.charCodeAt(0);
        if (code < 0x20 || code === 0x7f) {
          out += "\\x" + code.toString(16).padStart(2, "0");
        } else {
          out += ch;
        }
      }
    }
  }
  out += '"';
  return out;
}

// desugarConsList checks if a list is a cons chain:
// (cons a (cons b (cons c nil))) -> { elems: [a, b, c], tail: nil }
// (cons a (cons b tail))         -> { elems: [a, b], tail }
function desugarConsList(l: SList): { elems: Sexpr[]; tail: Sexpr } | null {
  if (l.elems.length !== 3 || !isSym(l.elems[0], "cons")) {
    return null;
  }

  const elems: Sexpr[] = [l.elems[1]];
  let rest: Sexpr = l.elems[2];

  while (true) {
    if (
      rest.kind !== "list" ||
      rest.elems.length !== 3 ||
      !isSym(rest.elems[0], "cons")
    ) {
      return { elems, tail: rest };
    }
    elems.push(rest.elems[1]);
    rest = rest.elems[2];
  }
}
